### PR TITLE
Add dynamic reward module

### LINKED ISSUE #7717: Add dynamic reward module
### Description

- Copy src and test for `reward` module and rename all parts to Dynamic module
- Update to use dynamic rewards in the default application in application.ts
- Update `beforeTransactionsExecute` and `afterTransactionsExecute` logics
  - it requires 2 new stores in the execute functions

### Acceptance Criteria

- Original reward module is kept as it is
- Default application uses dynamic reward module
- New logics are unit tested

### Additional Information
All the specification is defined in the proposed [LIP](https://research.lisk.com/t/introduce-dynamic-block-rewards-module/387)

### PR DESCRIPTION
**Pull request by shuse2**
### What was the problem?

This PR resolves #7717 

### How was it solved?

- Add new dynamic reward module using reward module
- Update genesis block for example app
- Add small refactor to reward module to be usable from dynamic reward module
- Add new DPoS method which is required by dynamic reward module
  - #7715 should add the implementation for `updateSharedRewards`
### How was it tested?

- Add unit tests for all the cases

### CODE REVIEW

File: framework/src/application.ts [hunk 1]
```diff
@@ -54,6 +53,7 @@ import { SidechainInteroperabilityModule } from './modules/interoperability/side
 import { MainchainInteroperabilityModule } from './modules/interoperability/mainchain/module';
 import { SidechainInteroperabilityMethod } from './modules/interoperability/sidechain/method';
 import { MainchainInteroperabilityMethod } from './modules/interoperability/mainchain/method';
```
Comment (mosmartin) [inline framework/src/application.ts:55]: Hi @shuse2 , this is not part of the PR but it's a small fix  😉. Maybe 
```ts
import {
	SidechainInteroperabilityModule,
	MainchainInteroperabilityModule,
	SidechainInteroperabilityMethod,
	MainchainInteroperabilityMethod,
} from './modules/interoperability';
```

Not a show stopper though.

**Review by mosmartin — APPROVED** [review]
🚀

File: framework/src/modules/dpos_v2/method.ts [hunk 2]
```diff
@@ -58,4 +64,29 @@ export class DPoSMethod extends BaseMethod {
...
+	public async updateSharedRewards(
+		_methodContext: MethodContext,
+		_generatorAddress: Buffer,
+		_tokenID: Buffer,
+		_reward: bigint,
+	): Promise<void> {
+		// TODO: Implement #7715
+	}
+
+	public isEndOfRound(_methodContext: ImmutableMethodContext, height: number): boolean {
```
Comment (bobanm) [inline framework/src/modules/dpos_v2/method.ts:85]: I think that `isLastBlockOfRound()` would be a more accurate name. The function can take any block height, from present, past or future. All it does is calculate if the given height belongs to the last block of the round.

To make my point even stronger, I saw that in `module.ts` you even assigned the result of this function to a variable named `isLastBlockOfRound` 🙂

Also, there is a much cleaner way to calculate the last block of the round, without using `Rounds` and creating all these new variables. It needs 1 line of code only:

```typescript
return height % this._config.roundLength === 0
```

That is if the first block has height 1. Otherwise, if the first block has height 0:

```typescript
return (height + 1) % this._config.roundLength === 0
```

Comment (bobanm) [reply framework/src/modules/dpos_v2/method.ts:85]: While continuing the code review, I noticed that it wasn't you who originally wrote this function. You just moved it from `DPoSModule` to `DPoSMethod`... and decided to rename it for some reason 🙂

Comment (shuse2) [reply framework/src/modules/dpos_v2/method.ts:85]: The name was defined in the [LIP](https://research.lisk.com/t/introduce-dynamic-block-rewards-module/387#after-transactions-execution-42) as `method`, so I used as it is.
The method itself has been used for a while, so i didn't change it.
round needs to consider genesis height, so the calculation should be 
```
(h - genesisDataStore.height) % ROUND_LENGTH == 0
```
Ill update the logic

File: framework/src/modules/reward/endpoint.ts [hunk 3]
```diff
@@ -15,17 +15,17 @@
 import { ModuleEndpointContext } from '../../types';
 import { BaseEndpoint } from '../base_endpoint';
 import { calculateDefaultReward } from './calculate_reward';
-import { DefaultReward, EndpointInitArgs } from './types';
+import { DefaultReward, ModuleConfig } from './types';
+
+export interface EndpointInitArgs {
+	config: ModuleConfig;
```
Comment (bobanm) [inline framework/src/modules/reward/endpoint.ts:21]: This `EndpointInitArgs` interface seems completely unnecessary. It doesn't add much value, as it only wraps `ModuleConfig` interface, without providing anything new.

How about we get rid of that boilerplate, and pass `ModuleConfig` object to `RewardEndpoint.init()`. That is how I implemented it recently in `TokenEndpoint`:

```typescript
public init(moduleConfig: ModuleConfig) {
    this._moduleConfig = moduleConfig;
}
```

File: framework/test/unit/modules/dynamic_rewards/module.spec.ts [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+				{ address: standbyValidatorAddress, bftWeight: BigInt(0) },
+			];
+
+			(validatorsMethod.getValidatorsParams as jest.Mock).mockResolvedValue({ validators });
+			(dposMethod.getNumberOfActiveDelegates as jest.Mock).mockReturnValue(activeDelegate);
+		});
+
+		it('should store minimal reward for active delegates when full round is forged', async () => {
+			// Round is already completed once
+			const generatorMap = new Array(103).fill(0).reduce(prev => {
```
Comment (bobanm) [inline framework/test/unit/modules/dynamic_rewards/module.spec.ts:193]: Define a constant such as `ROUND_LENGTH` instead of using a literal integer 103. Both here and when initializing `dposMethod`. It makes it clearer what 103 means, and it's easier to maintain if we ever want to test with different round lengths.

File: framework/test/unit/modules/dynamic_rewards/module.spec.ts [hunk 5]
```diff
@@ -47,6 +47,9 @@ import {
 } from '../../../../src/modules/reward/constants';
 
 describe('DynamicRewardModule', () => {
+	const defaultRoundLength = 103;
+	const defaultNumberOfActiveDelegates = 101;
```
Comment (bobanm) [inline framework/test/unit/modules/dynamic_rewards/module.spec.ts:51]: What's the convention about read only params constants? Shouldn't we use UPPERCASE? 🤔

Comment (shuse2) [reply framework/test/unit/modules/dynamic_rewards/module.spec.ts:51]: it's a constant (variable) within a function, so i think it shouldnt be uppercase

**Review by bobanm — APPROVED** [review]
